### ISSUE TITLE
Some issues on boundary setting in MultiSlice algorithm

### ISSUE THREAD

**Issue by SoverHHH** [abTEM/abTEM#166]
Hello,

I am trying to use abTEM to calculate a STEM image of a sample and have encountered some issues. I obtained the atomic positions of a sample using the AET method. In the usual workflow, after getting all the atomic positions, we typically use the MultiSlice algorithm for simulation and then perform a reconstruction using the simulated image to compare with the original reconstruction to estimate the accuracy of the results.

My sample contains 80,000 atoms. Considering computational efficiency, I divided the sample into two halves (each with 40,000 atoms) for calculation. I also used the PRISM algorithm provided by abTEM to speed up the calculation. However, after performing a series of tests, I noticed some strange phenomena. Below are the tests I conducted.

First, I obtained a part B (61.8 Å), then in each test, I removed the 0-10 Å section in the y-dimension and recalculated, obtaining sb (51.8 Å), 2sb (41.8 Å), and 3sb (31.8 Å). In each test, I used `s_matrix.dummy_probes` to check for probe self-interference and displayed the final results.

- Part B, thickness 61.8 Å, no self-interference, image relatively clear, took 111 min on a 3090
![image](https://github.com/user-attachments/assets/26e231bc-369e-4c82-95e9-e1953ebfb61f)
![image](https://github.com/user-attachments/assets/ccc7817e-3e2c-4001-847f-30fa4f23c70c)

- Sb, thickness 51.8 Å, slight self-interference, image less clear than Part B, took 96 min on a 3090
![image](https://github.com/user-attachments/assets/22fbcc79-d607-4fe6-b2a1-471f55e3bcfa)
![image](https://github.com/user-attachments/assets/55820e46-8b48-439a-adf6-3ff8367224f8)

- 2sb, thickness 41.8 Å, significant self-interference, simulated image very blurry
![image](https://github.com/user-attachments/assets/374e6d7a-f32b-4deb-86ce-2fa1a799ad01)
![image](https://github.com/user-attachments/assets/bad24dbb-4e0b-4b31-ad60-9a15d2a0d150)

- 3sb, thickness 31.8 Å, significant self-interference, but the simulated image in the same area is much clearer than 2sb, took 28 min on a 3070
![image](https://github.com/user-attachments/assets/59a5f698-0ff7-4817-bbd7-c9ea55da4e7c)
![image](https://github.com/user-attachments/assets/6f225106-88bf-4390-b65e-97045c8ed5a6)

From these tests, it can be observed that the smaller the thickness, the more significant the probe self-interference. From Part B to 2sb, the simulated image becomes blurrier, but the STEM image of 3sb appears very clear. After carefully reading the abTEM documentation, I found that the program uses periodic boundary conditions by default in the x and y directions when calculating images. I manually specified the `atoms.cell` parameter after importing the atomic coordinates from the XYZ file, and I suspect this might cause some issues at the periodic boundaries.

```python
atoms = ase.io.read(folder + '/partb.xyz')
Mycell = [0, 0, 0]
for i in range(3):
    Mycell[i] = max(atoms.positions[:, i]) - min(atoms.positions[:, i])
atoms.cell = Mycell
```

I used the above code to obtain the cell dim parameters and then divided by the lattice constant in the y direction (3.94 Å) to get the following values:

| y-direction thickness | y-direction lattice number |
|-----------------------|----------------------------|
| Part B 61.8 Å         | 15.6853                    |
| Sb 51.8 Å             | 13.1472                    |
| 2sb 41.8 Å            | 10.6091                    |
| 3sb 31.8 Å            | 8.0711                     |

It can be seen that the clarity of the 3sb image may be related to the cell value being close to an integer multiple of the lattice constant.

Therefore, my questions can be summarized as follows:
1. Does setting the cell dim as an integer multiple of the lattice constant improve image quality?
2. My sample does not actually have periodic boundary conditions. Is there a way to adjust certain parameters to obtain a more accurate electron microscope simulation? For example, by adding some blank space around the sample?

I would appreciate your advice.

Below is the code I used:
```py
import matplotlib.pyplot as plt
import ase
import h5py
import dask
import abtem
from ase.build import graphene
from ase.io import write
abtem.config.set({"device": "gpu"})
dask.config.set({"num_workers": 1})
abtem.config.set({"dask.chunk-size-gpu" : "512MB"})


atoms=ase.io.read('./partb_13.xyz')
Mycell=[0,0,0]
for i in range(0,3):
    Mycell[i]=max(atoms.positions[:,i])-min(atoms.positions[:,i])
atoms.cell=Mycell

potential = abtem.Potential(
    atoms,
    sampling=0.05,
    parametrization="lobato",
    # slice_thickness=(max(atoms.positions[:,2])-min(atoms.positions[:,2]))/100,
    slice_thickness=2,
    projection="finite",
)
probe = abtem.Probe(energy=300e3, semiangle_cutoff=25, defocus="scherzer")
probe.grid.match(potential)

grid_scan = abtem.GridScan(
    start=(0, 0),
    end=(1, 1),
    sampling=0.3079,
    fractional=True,
    potential=potential,
)
s_matrix = abtem.SMatrix(
    potential=potential, energy=300e3, semiangle_cutoff=25, interpolation=4,store_on_host=True
)
Cs = 8e-6 * 1e10  # 20 micrometers
ctf = abtem.CTF(Cs=Cs, defocus="scherzer", energy=s_matrix.energy)
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(20, 20))
s_matrix.dummy_probes(plane="entrance", ctf=ctf).show(ax=ax1, title="entrance", power=1)
s_matrix.dummy_probes(plane="exit", ctf=ctf).show(ax=ax2, title="exit", power=1);

detector = abtem.FlexibleAnnularDetector()
flexible_measurement = s_matrix.scan(scan=grid_scan,detectors=detector,ctf=ctf,max_batch_multislice = 30, max_batch_reduction = 300)
flexible_measurement.compute()

bf_measurement = flexible_measurement.integrate_radial(0, probe.semiangle_cutoff)
maadf_measurement = flexible_measurement.integrate_radial(50, 129)
maadf_measurement.show(
    explode=True,
    figsize=(20, 20),
    cbar=True,
);
```

**Comment by TomaSusi** [comment]
Hi! Thanks for providing the code. It's a bit of a complicated situation with a custom atomic structure and PRISM, but I have a few initial thoughts.

1. I'm a bit confused by you using the word "thickness" to describe the y-extent of the model. The propagation direction in abTEM is the positive z-direction, so thickness in our usage means the z-thickness; y could be called "height" here. Are we on the same page about this and this was just semantics?

2. Integer multiples of cell in the x-y directions should play no role in the result; in the propagation z direction there may be some benefit to divide the cell nicely.

3. As for non-periodic boundaries, you can always add vacuum at the sides to emulate that. After setting the cell of the atoms object as `Mycell`, center the model with the command `atoms.center(10, axis=(0,1))` to get 10 A of vacuum in along axes 0 and 1 (ie. x and y).

4. We do not recommend splitting your model into parts precisely because of periodicity; you cannot stitch such simulations together to get the same result as simulating the entire model. 80 000 atoms should still be quite doable with PRISM, and there is one important "trick" you can use to significantly reduce memory use, if that is a bottleneck for you.

Namely, instead of `flexible_measurement.compute()`, you can directly compute the results to disk in chunks by using `flexible_measurement.to_zarr(filename)`, and then read them back in lazily with flexible_measurements = abtem.from_zarr(filename)`. This way the entire measurement never needs to necessarily fit in memory.

**Comment by SoverHHH** [comment]
@TomaSusi 
Thank you very much for your response! I carefully read through your reply, modified my code, and ran another test.

1. "Thickness" is indeed a semantic issue because, during testing, I need to extract samples with different thicknesses in the y-direction. I imagined it as slicing a piece of bread, so I used "thickness" to represent the thickness in the y-direction. I agree with your point that "thickness" is usually used to refer to the z-direction, and I will correct this usage going forward.

2. My local device is a Dell Windows workstation running Windows 11, with 512GB of memory and an RTX 3090 GPU with 24GB of VRAM. Additionally, Windows 11 can convert 255GB of system memory into shared GPU memory. When running a 40k-atom model with the parameters mentioned above (with GPU acceleration), the total GPU memory usage reaches over 200GB. Unsurprisingly, when computing the full 80k-atom model, an out-of-memory error occurred. More specifically, the PRISM algorithm seems to have two stages in terms of memory usage. For the 40k-atom case, the GPU memory usage is around 23GB in the first stage, but it reaches over 200GB in the second stage.
![image](https://github.com/user-attachments/assets/6d066b42-96ac-43c2-80f4-f43b88111ed8)

The MultiSlice algorithm doesn't have memory overflow issues, but it's significantly slower compared to PRISM. In previous tests, the same task took tens of times longer. Since I need to compute 70 electron microscopy images at different rotation angles, the time required is just too long. I believe that more efficient electron microscopy simulation is achievable, and I wonder if you are interested in further exploring this direction.

3. Following your suggestion, I used 'flexible_measurement.to_zarr('0902test13')' in my code to replace the original 'flexible_measurement.compute()'. However, from the results, there doesn't seem to be much improvement. A new folder named '0902test13' was created in the current path, and after running the 40k-atom model, its size was 64.5MB (I didn't monitor changes during runtime). This size is similar to the results stored in the h5 format, but it's several orders of magnitude smaller than the 200+GB I mentioned earlier. I’m unsure if I’m using the method correctly.

I look forward to more of your suggestions.

**Comment by TomaSusi** [comment]
2. I am not an expert on the PRISM algorithm, but it may indeed need more memory than our multislice implementation. Have you tried our hints for managing memory use in PRISM?
https://abtem.github.io/doc/user_guide/tutorials/prism.html#memory-issues

3. The point of computing directly to disk is that the total amount of memory needed at any time can be much smaller, though the output files that are generated will be exactly the same size. Is your maximum memory usage still so high with PRISM when using `to_zarr()`? This approach may be more optimal for multislice since that is can be more straightforwardly divided into independent chunks.

**Comment by TomaSusi** [comment]
Btw, since this is a scanned measurement, have you tried to only scan smaller parts of the potential at a time? Combining separate but connected scan regions will be possible since at the edge of each area the probe still sees the potential beyond as normal. Not sure how much memory saving this will give in PRISM, but it is much better justified than cutting up your specimen.

**Comment by SoverHHH** [comment]
@TomaSusi 
Thank you very much for your response again! Based on your suggestions, I adjusted some parameters and tested again. The memory usage has significantly decreased in the early stages, but the speed also slowed down considerably, which made the testing take more time. Below are the details of the situation.

1. I have read the hints you provided on the PRISM page and tried the following adjustments in different functions:
   `max_batch_multislice=10, max_batch_reductionu=100, store_on_host=True`. Additionally, I reduced the scanning range to `start=(0,0), end=(0.5,0.5)`. This time, after approximately 10 hours of running, I encountered an out-of-memory error on the GPU.

2. I further significantly reduced the parameters to `max_batch_multislice=5, max_batch_reductionu=50, store_on_host=True`. In the first stage, the GPU memory usage was only around 5GB, which is about half of the original memory usage. The GPU utilization was over 90% on average. After 20 hours, I checked and saw the memory usage remained consistent. However, after 40 hours, I encountered another out-of-memory error. The error message showed that the program attempted to allocate 424GB to the GPU memory, which is clearly impossible.
![image](https://github.com/user-attachments/assets/274aa784-5e4e-4c48-b78a-d7f4b203d9fe)
Additionally, on your PRISM Memory Issues page, there is a section about frozen phonons. However, I believe I did not use it during my run (you can refer to my code above), because based on my understanding, different frozen phonons mean repeating the calculation multiple times to improve the simulation accuracy. However, for my model, a single calculation already takes enough time, so I did not set it up.

In summary, it seems that if I still want to use abtem to get the electron microscopy simulation images, I might need to:
1. Use the MultiSlice algorithm, although it will be slower.
2. Lower the simulation accuracy and continue using the PRISM algorithm.
3. Split the model into two halves and calculate them separately, although as you mentioned, this is not equivalent to directly simulating the full model.

Thank you very much for your help!

**Comment by TomaSusi** [comment]
Correct, you can ignore the frozen phonons in your context.

Did you try to use the full model but only a small scan area? Is the memory use the same?

Thanks for testing the memory management options, based on that massive spike in usage at the end of the calculation, it suggests to me that something is fundamentally memory-limited with the PRISM algorithm, or that we have a bug in our memory management. Unfortunately I'm not sure we will be able to work on improving this in the near future, your system is kind of extreme.

You could try running the same system using Prismatic to see if that behaves similarly, but I cannot help you more with that. An alternative using abTEM, if you have access to compute nodes with multiple GPUs, is to do multi-GPU parallelization of the multislice algorithm, which should give a nice speedup.

**Comment by TomaSusi** [comment]
One more idea courtesy of the Luis on the NCEM Slack: could you try `flexible_measurement.to_zarr(compute=False)`? This should be default behavior but maybe there is a bug related to PRISM. Alternatively, you could try pushing up the interpolation factor.

**Comment by SoverHHH** [comment]
I'm sorry, I won't have the time to try your suggestion again until later. Thank you for your response.

**Comment by TomaSusi** [comment]
This PR is likely to help: https://github.com/abTEM/abTEM/pull/269
